streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -146,96 +165,203 @@ public void processInOrder(final K key, final V value, final long timestamp) {
+                    key,
+                    key,
+                    0,
+                    // to catch the current record's right window, if it exists, without more calls to the store
+                    inputRecordTimestamp + 1)
+            ) {
+                while (iterator.hasNext()) {
+                    final KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> windowBeingProcessed = iterator.next();
+                    final long startTime = windowBeingProcessed.key.window().start();
+                    windowStartTimes.add(startTime);
+                    final long windowMaxRecordTimestamp = windowBeingProcessed.value.timestamp();
+
+                    if (startTime == 0) {
+                        combinedWindow = windowBeingProcessed;
+                        //We don't need to store previousRecordTimestamp if maxRecordTimestamp > timestamp
+                        // because the previous record's right window (if there is a previous record)
+                        // would have already been created by maxRecordTimestamp
+                        if (windowMaxRecordTimestamp < inputRecordTimestamp) {
+                            previousRecordTimestamp = windowMaxRecordTimestamp;
+                        }
+
+                    } else if (startTime <= inputRecordTimestamp) {
+                        rightWinAgg = windowBeingProcessed.value;
+                        updateWindowAndForward(windowBeingProcessed.key.window(), windowBeingProcessed.value, key, value, closeTime, inputRecordTimestamp);
+                    } else if (startTime == inputRecordTimestamp + 1) {
+                        rightWinAlreadyCreated = true;
+                    } else {
+                        throw new IllegalStateException("Unexpected window found when processing sliding windows");
+                    }
+                }
+            }
+
+            // If there wasn't a right window agg found and we need a right window for our new record,
+            // the current aggregate in the combined window will go in the new record's right window. We can be sure that the combined
+            // window only holds records that fall into the current record's right window for two reasons:
+            // 1. If there were records earlier than the current record AND later than the current record, there would be a right window found
+            // when we looked for right window agg.
+            // 2. If there was only a record before the current record, we wouldn't need a right window for the current record and wouldn't update the
+            // rightWinAgg value here, as the combinedWindow.value.timestamp() < inputRecordTimestamp
+            if (rightWinAgg == null && combinedWindow != null && combinedWindow.value.timestamp() > inputRecordTimestamp) {
+                rightWinAgg = combinedWindow.value;
+            }
+
+            //create right window for new record if needed
```

Comment:
nit: this code is pretty self-explanatory now, so the comment isn't really adding anything